File: src/utils/scan_map.c
//
// Created by anerruption on 06/03/23.
//

#include "../../include/utils/scan_map.h"

key get_keyboard_key(uint8_t scan_code)
{
    key key;

    switch (scan_code)
    {
        case 14:
            key.type = BACKSPACE;
            break;
        case 16:
            key.type = NORMAL;
            key.value = 'q';
            break;
        case 17:
            key.type = NORMAL;
            key.value = 'w';
            break;
        case 18:
            key.type = NORMAL;
            key.value = 'e';
            break;
        case 19:
            key.type = NORMAL;
            key.value = 'r';
            break;
        case 20:
            key.type = NORMAL;
            key.value = 't';
            break;
        case 21:
            key.type = NORMAL;
            key.value = 'y';
            break;
        case 22:
            key.type = NORMAL;
            key.value = 'u';
            break;
        case 23:
            key.type = NORMAL;
            key.value = 'i';
            break;
        case 24:
            key.type = NORMAL;
            key.value = 'o';
            break;
        case 25:
            key.type = NORMAL;
            key.value = 'p';
            break;
        case 28:
            key.type = ENTER;
            break;
        case 30:
            key.type = NORMAL;
            key.value = 'a';
            break;
        case 31:
            key.type = NORMAL;
            key.value = 's';
            break;
        case 32:
            key.type = NORMAL;
            key.value = 'd';
            break;
        case 33:
            key.type = NORMAL;
            key.value = 'f';
            break;
        case 34:
            key.type = NORMAL;
            key.value = 'g';
            break;
        case 35:
            key.type = NORMAL;
            key.value = 'h';
            break;
        case 36:
            key.type = NORMAL;
            key.value = 'j';
            break;
        case 37:
            key.type = NORMAL;
            key.value = 'k';
            break;
        case 38:
            key.type = NORMAL;
            key.value = 'l';
            break;
        case 44:
            key.type = NORMAL;
            key.value = 'z';
            break;
        case 45:
            key.type = NORMAL;
            key.value = 'x';
            break;
        case 46:
            key.type = NORMAL;
            key.value = 'c';
            break;
        case 47:
            key.type = NORMAL;
            key.value = 'v';
            break;
        case 48:
            key.type = NORMAL;
            key.value = 'b';
            break;
        case 49:
            key.type = NORMAL;
            key.value = 'n';
            break;
        case 50:
            key.type = NORMAL;
            key.value = 'm';
            break;
        case 57:
            key.type = NORMAL;
            key.value = ' ';
            break;
        default:
            key.type = UNKNOWN;
            break;
    }

    return key;
}

File: src/utils/scanmap.zig
pub const KeyType = enum { unknown, normal, enter, backspace };
pub const Key = struct { type: KeyType, value: u8 };

pub fn getKey(scan_code: u8) Key {
    return switch (scan_code) {
        14 => .{ .type = .backspace, .value = 0 },
        16 => .{ .type = .normal, .value = 'q' },
        17 => .{ .type = .normal, .value = 'w' },
        18 => .{ .type = .normal, .value = 'e' },
        19 => .{ .type = .normal, .value = 'r' },
        20 => .{ .type = .normal, .value = 't' },
        21 => .{ .type = .normal, .value = 'y' },
        22 => .{ .type = .normal, .value = 'u' },
        23 => .{ .type = .normal, .value = 'i' },
        24 => .{ .type = .normal, .value = 'o' },
        25 => .{ .type = .normal, .value = 'p' },
        28 => .{ .type = .enter, .value = 0 },
        30 => .{ .type = .normal, .value = 'a' },
        31 => .{ .type = .normal, .value = 's' },
        32 => .{ .type = .normal, .value = 'd' },
        33 => .{ .type = .normal, .value = 'f' },
        34 => .{ .type = .normal, .value = 'g' },
        35 => .{ .type = .normal, .value = 'h' },
        36 => .{ .type = .normal, .value = 'j' },
        37 => .{ .type = .normal, .value = 'k' },
        38 => .{ .type = .normal, .value = 'l' },
        44 => .{ .type = .normal, .value = 'z' },
        45 => .{ .type = .normal, .value = 'x' },
        46 => .{ .type = .normal, .value = 'c' },
        47 => .{ .type = .normal, .value = 'v' },
        48 => .{ .type = .normal, .value = 'b' },
        49 => .{ .type = .normal, .value = 'n' },
        50 => .{ .type = .normal, .value = 'm' },
        57 => .{ .type = .normal, .value = ' ' },
        else => .{ .type = .unknown, .value = 0 },
    };
}
